build_docs_server.py

```diff
@@ -0,0 +1,142 @@
+import asyncio
+import argparse
+import logging.config
+import os
+
+from aiohttp import web
+import yaml
+
+__version__ = "0.0.1"
+
+DEFAULT_LOGGING_CONFIG = """
+---
+
+version: 1
+disable_existing_loggers: false
+formatters:
+  normal:
+    format: '%(asctime)s - %(levelname)s - %(message)s'
+handlers:
+  stderr:
+    class: logging.StreamHandler
+    stream: ext://sys.stderr
+    level: DEBUG
+    formatter: normal
+loggers:
+  build_docs_server:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.access:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.client:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.internal:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.server:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.web:
+    level: DEBUG
+    handlers: [stderr]
+  aiohttp.websocket:
+    level: DEBUG
+    handlers: [stderr]
+"""
+
+logger = logging.getLogger("build_docs_server")
+
+
+async def version(request):
+    return web.json_response(
+        {
+            "name": "docs.python.org Github handler",
+            "version": __version__,
+            "source": "https://github.com/python/docsbuild-scripts",
+        }
+    )
+
+
+async def child_waiter(app):
+    while True:
+        try:
+            status = os.waitid(os.P_ALL, 0, os.WNOHANG | os.WEXITED)
+            logger.debug("Child completed with status %s", str(status))
+        except ChildProcessError:
+            await asyncio.sleep(600)
+
+
+async def start_child_waiter(app):
+    app["child_waiter"] = asyncio.ensure_future(child_waiter(app))
+
+
+async def stop_child_waiter(app):
+    app["child_waiter"].cancel()
+
+
+async def hook(request):
+    payload = await request.json()
```

Comment:
We will probably want to validate GitHub hooks.

Here is sample code from another project (so it's not drop in... just shows the mechanism)

```python
    def validate_webhook(self):
        if self.webhook_secret is None:
            return True
        return hmac.compare_digest(
            request.headers.get('X-Hub-Signature').split('=')[1],
            hmac.new(self.webhook_secret.encode(), msg=request.data, digestmod=hashlib.sha1).hexdigest()
        )
```

This will also require an agreed on secret that can be configured via salt.

Reply:
Done. I bet we do not even need to agree on the secret: as I don't have the rights to add the webhook on cpython you'll be the one configuring salt side and cpython repo side :p